internal/cache/async.go

```diff
@@ -103,49 +103,41 @@ func (ac *asyncClient) doUpdate(ctx context.Context, key store.Key) {
 	switch {
 	case err == nil:
 		ac.objectStore.OverrideResourceVersionIfNewer(result)
-	case isRetryableError(err):
-		svc1log.FromContext(ctx).Warn("got retryable error, will retry", svc1log.Stacktrace(err))
-		ac.queue.AddIfAbsent(store.UpdateRequest(obj))
 	case errors.IsConflict(err):
 		svc1log.FromContext(ctx).Warn("got conflict, will try updating resource version", svc1log.Stacktrace(err))
-		newObj, getErr := ac.client.Get(key.Namespace, key.Name)
+		newObj, getErr := ac.client.Get(r.Key.Namespace, r.Key.Name)
 		switch {
 		case getErr == nil:
 			ac.objectStore.OverrideResourceVersionIfNewer(newObj)
-			ac.doUpdate(ctx, key)
-		case isRetryableError(getErr):
-			svc1log.FromContext(ctx).Warn("got retryable error, will retry", svc1log.Stacktrace(getErr))
-			ac.queue.AddIfAbsent(store.UpdateRequest(obj))
+			ac.doUpdate(ctx, store.UpdateRequest(newObj))
 		default:
-			logNonRetryableError(ctx, err)
+			ac.maybeRetryRequest(ctx, r, getErr)
 		}
 	default:
-		logNonRetryableError(ctx, err)
+		ac.maybeRetryRequest(ctx, r, err)
 	}
 }
 
-func (ac *asyncClient) doDelete(ctx context.Context, key store.Key) {
-	err := ac.client.Delete(key.Namespace, key.Name)
+func (ac *asyncClient) doDelete(ctx context.Context, r store.Request) {
+	err := ac.client.Delete(r.Key.Namespace, r.Key.Name)
 	switch {
 	case err == nil:
 		return
-	case isRetryableError(err):
-		svc1log.FromContext(ctx).Warn("got retryable error, will retry", svc1log.Stacktrace(err))
-		ac.queue.AddIfAbsent(store.Request{Key: key, Type: store.DeleteRequestType})
 	case errors.IsNotFound(err):
 		svc1log.FromContext(ctx).Info("object already deleted")
 	default:
-		logNonRetryableError(ctx, err)
+		ac.maybeRetryRequest(ctx, r, err)
 	}
 }
 
-func isRetryableError(err error) bool {
-	return errors.IsServerTimeout(err) || errors.IsServiceUnavailable(err) ||
-		errors.IsTooManyRequests(err) || errors.IsTimeout(err)
-}
-
-func logNonRetryableError(ctx context.Context, err error) {
-	svc1log.FromContext(ctx).Error("got non retryable error", svc1log.Stacktrace(err))
+func (ac *asyncClient) maybeRetryRequest(ctx context.Context, r store.Request, err error) bool {
+	if r.RetryCount >= ac.config.MaxRetryCount() {
+		svc1log.FromContext(ctx).Error("max retry count reached, dropping request", svc1log.Stacktrace(err))
+		return false
+	}
+	svc1log.FromContext(ctx).Warn("got retryable error, will retry", svc1log.Stacktrace(err))
+	ac.queue.AddIfAbsent(r.WithIncrementedRetryCount())
```

Comment:
why addIfAsbent here? What is the case where it would not be absent in the queue?

Reply:
A different client of the cache could have added a request to update the object (for example) while we were processing this one.